rust/noosphere-gateway/src/middleware.rs

```diff
@@ -0,0 +1,244 @@
+    ) -> Result<(), StatusCode> {
+        let capability_infos = self.proof_chain.reduce_capabilities(&SPHERE_SEMANTICS);
+
+        for capability_info in capability_infos {
+            trace!("Checking capability: {:?}", capability_info.capability);
+            if capability_info
+                .originators
+                .contains(self.counterpart().as_str())
+                && capability_info.capability.enables(capability)
+            {
+                debug!("Authorized!");
+                return Ok(());
+            }
+        }
+
+        Err(StatusCode::UNAUTHORIZED)
+    }
+
+    pub fn sphere(&self) -> &C {
+        self.sphere.sphere()
+    }
+
+    pub fn sphere_mut(&mut self) -> &mut C {
+        self.sphere.sphere_mut()
+    }
+
+    pub fn counterpart(&self) -> &Did {
+        self.sphere.counterpart()
+    }
+
+    pub async fn sphere_context(&self) -> Result<C::SphereContext> {
+        self.sphere.sphere_context().await
+    }
+
+    pub async fn sphere_context_mut(&mut self) -> Result<C::MutableSphereContext> {
+        self.sphere.sphere_context_mut().await
+    }
+}
+
+#[async_trait]
+impl<M, C, S, State> FromRequestParts<State> for GatewayAuthority<M, C, S>
+where
+    State: Send + Sync,
+    M: ContextManager<C, S> + 'static,
+    C: HasMutableSphereContext<S>,
+    S: Storage + 'static,
+{
+    type Rejection = StatusCode;
+
+    async fn from_request_parts(parts: &mut Parts, state: &State) -> Result<Self, Self::Rejection> {
+        let sphere = GatewaySphereContext::from_request_parts(parts, state)
+            .await
+            .map_err(map_bad_request)?;
+
+        let TypedHeader(Authorization(bearer)) =
+            TypedHeader::<Authorization<Bearer>>::from_request_parts(parts, state)
+                .await
+                .map_err(map_bad_request)?;
+
+        let ucans: noosphere_headers::Ucan =
+            TypedHeader::<noosphere_headers::Ucan>::from_request_parts(parts, state)
+                .await
+                .map_err(map_bad_request)?
+                .0;
+
+        let db = {
+            let sphere_context: C::SphereContext =
+                sphere.sphere_context().await.map_err(map_bad_request)?;
+            sphere_context.db().clone()
+        };
+
+        let mut did_parser = DidParser::new(SUPPORTED_KEYS);
+        let proof_chain = ucans
+            .as_proof_chain(bearer, db, &mut did_parser)
+            .await
+            .map_err(map_bad_request)?;
+
+        Ok(GatewayAuthority {
+            proof_chain,
+            sphere,
+        })
+    }
+}
+
+fn map_bad_request<E: std::fmt::Debug>(error: E) -> StatusCode {
+    error!("{:?}", error);
+    StatusCode::BAD_REQUEST
+}
+
+/// Takes arguments from a [FromRequestParts::from_request_parts] call and parses out
+/// the counterpart sphere identity from hostname.
+/// e.g. `did_key_abc.subconscious.cloud` -> `Did(String::from("did_key_abc"))`
+async fn get_counterpart_from_hostname<State>(
```

Comment:
This needs some careful thought, because pulling the counterpart out of the hostname is not something we want to formalize as a "Noosphere Protocol" detail. I would be more comfortable if a "counterpart resolution" function was given to the gateway at the time it is instantiated.

Reply:
My understanding was that this *is* a part of the protocol, due to secure origins on the web. Exposing this (currently, without a more modular V2 that exposes individual routes where a consumer BYO's axum) with axum guts may get weird. WDYT if we had two options for counterpart parsing, one as a subdomain, one as a header or query param, not dissimilar from IPFS gateways, such that the implementations live with the other axum stuff for now?

Reply:
> My understanding was that this is a part of the protocol, due to secure origins on the web

Nope, this is a requirement for our cloud if we intend to serve a client over the web, because we'll be serving multiple customers and we don't want their origins to cross over.

A self-hosting individual user doesn't have the same requirement.

Reply:
Addressed!